File: emr_mine_python_scipts/pq_tree/common_intervals.py
'''
Created on Dec 29, 2010

@author: patnaik
'''


#def naive_common_interval(pi_AB, C_prev = None):
#    n = len(pi_AB)
#    C = set()
#    for x in xrange(n-1):
#        l = u = pi_AB[x]
#        for y in xrange(x+1, n):
#            l = min(l, pi_AB[y])
#            u = max(u, pi_AB[y])
#            if u - l - (y - x) == 0 and (C_prev == None or (l,u) in C_prev):
#                C.add((l,u))
#    return C


def naive_common_interval_order(pi_AB, pos_map, C_prev):
    n = len(pi_AB)
    C = set()
    for x in xrange(n):
        l = u = pi_AB[x]
        for y in xrange(x, n):
            l = min(l, pi_AB[y])
            u = max(u, pi_AB[y])
            if (u - l) - (y - x) == 0 and (C_prev == None or (l,u) in C_prev):
                pos_map[(l,u)].append(x)
                C.add((l,u))
    return C


from pygraph.classes.digraph import digraph
from pygraph.algorithms.critical import transitive_edges



def reducegraph(edges):
    g = digraph()
    nodes = set()
    for (x, y) in edges:
        if x not in nodes:
            nodes.add(x)
            g.add_node(x)
        if y not in nodes:
            nodes.add(y)
            g.add_node(y)
        g.add_edge((x,y))

    for (x,y) in transitive_edges(g):
        if (x,y) in edges: edges.remove((x,y))
    return edges



def make_tree(C, orders):
    node_map = dict([(tuple(C[k]), k) for k in xrange(len(C))])
    #print node_map
    nodef = lambda x: node_map[C[x]]
    edges = set()
    for i in xrange(len(C)):
        for j in xrange(len(C)):
            if i == j: continue
            if C[i][0] <= C[j][0] and C[i][1] >= C[j][1]:
                edges.add((nodef(i), nodef(j)))

    edges = reducegraph(edges)
    edges = list(edges)
    edges.sort()
    for (i, j) in edges:
        print C[i], "->", C[j]
    

def inv_map(seq_1, seq_2): #O(n)
    m = {}
    for i in xrange(len(seq_1)):
        m[seq_1[i]] = i
    return [m[x] for x in seq_2]
    

def common_intervals(sequences, pos_map):
    C = None
    for i in xrange(0,len(sequences)):
        pi_AB = inv_map(sequences[0], sequences[i])
        C = naive_common_interval_order(pi_AB, pos_map, C)
        #C = naive_common_interval(pi_AB, C)
    if C == None: C = []
    else:
        C = list(C)
        C.sort()
    return C
    
class Node(object):
    def __init__(self):
        self.parent = None
        self.value = 0
        self.type = 'L'
        self.children = []



from collections import defaultdict
if __name__ == '__main__':
    print 'Common interval'
    
    sequences = [(1,2,3,4,5,6,7,8,9), (9,8,4,5,6,7,1,2,3), (1,2,3,8,7,4,5,6,9)]
    #sequences = [(1,2,3,4,5,6), (3,1,2,4,5,6), (2,3,1,4,5,6)]
    #sequences = [(1,2,3), (2,3,1), (3,2,1)]
    #sequences = [(1,2,3,4), (2,1,3,4), (1,2,4,3), (2,1,4,3)]
    #sequences = [(1,2,3,4,5), (1,2,3,4,5)]
    
    pos_map = defaultdict(list)
    C = common_intervals(sequences, pos_map)
    for (l,u) in C:
        print "(%d,%d)" % (l, u),
        print " locations =", pos_map[(l,u)]
    
    #make_tree(C, pos_map)
    
    
    

File: EMRVisualization-java-UI-tool/src/emr/partialorder/CommonIntervals.java
/*

 *  Copyright 2011 patnaik.

 */

/*
 * CommonIntervals.java
 *
 * Created on Jan 27, 2011, 6:35:43 PM
 */
package emr.partialorder;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;

/**
 *
 * @author patnaik
 */
public class CommonIntervals {

    public static HashSet<Interval> naive_common_interval(int[] pi_AB,
            HashMap<Interval, ArrayList<Integer>> pos_map,
            HashSet<Interval> C_prev) {
        int n = pi_AB.length;
        HashSet<Interval> C = new HashSet<Interval>();
        int l, u;
        for (int x = 0; x < n; x++) {
            l = u = pi_AB[x];
            for (int y = x; y < n; y++) {
                l = Math.min(l, pi_AB[y]);
                u = Math.max(u, pi_AB[y]);
                Interval ivl = new Interval(l, u);
                if (u - l - (y - x) == 0 && (C_prev == null || C_prev.contains(ivl))) {
                    if (!pos_map.containsKey(ivl)) {
                        pos_map.put(ivl, new ArrayList<Integer>());
                    }
                    pos_map.get(ivl).add(x);
                    C.add(ivl);
                }
            }
        }
        return C;
    }





    public static int[] inv_map(String[] seq_1, String[] seq_2) {
        HashMap<String, Integer> m = new HashMap<String, Integer>();
        for (int i = 0; i < seq_1.length; i++) {
            m.put(seq_1[i], i);
        }
        int[] retval = new int[seq_1.length];
        for (int i = 0; i < seq_1.length; i++) {
            retval[i] = m.get(seq_2[i]);
        }
        return retval;
    }


    public static ArrayList<Interval> common_intervals(ArrayList<String[]> sequences,
            HashMap<Interval, ArrayList<Integer>> pos_map) {
        
        int count = sequences.size();
        return common_intervals(sequences, pos_map, count);
    }


    public static ArrayList<Interval> common_intervals(ArrayList<String[]> sequences,
            HashMap<Interval, ArrayList<Integer>> pos_map, int count) {
        HashSet<Interval> C = null;
        ArrayList<Interval> C_list = new ArrayList<Interval>();
        for (int i = 0; i < count; i++) {
            int[] pi_AB = inv_map(sequences.get(0), sequences.get(i));
            C = naive_common_interval(pi_AB, pos_map, C);
        }
        if (C != null) {
            for (Interval ivl : C) {
                C_list.add(ivl);
            }
            Collections.sort(C_list);
        }
        return C_list;
    }
}
